File: C++/Easy/Accepted/flood-fill_7.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        
        int n = image.size();
        int m = image[0].size();
        vector<vector<bool>> vis(n, vector<bool>(m, false));
        int old_color = image[sr][sc];
        queue<pair<int, int>> q;
        q.push(make_pair(sr, sc));
        vis[sr][sc] = true;
        while(!q.empty()) {
            pair<int, int> p = q.front();
            q.pop();
            int x = p.first;
            int y = p.second;
            image[x][y] = color;
            if(x > 0 && image[x - 1][y] == old_color && !vis[x - 1][y]) {
                q.push(make_pair(x - 1, y));
                vis[x - 1][y] = true;
            }
            if(x < n - 1 && image[x + 1][y] == old_color && !vis[x + 1][y]) {
                q.push(make_pair(x + 1, y));
                vis[x + 1][y] = true;
            }
            if(y > 0 && image[x][y - 1] == old_color && !vis[x][y - 1]) {
                q.push(make_pair(x, y - 1));
                vis[x][y - 1] = true;
            }
            if(y < m - 1 && image[x][y + 1] == old_color && !vis[x][y + 1]) {
                q.push(make_pair(x, y + 1));
                vis[x][y + 1] = true;
            }
        }
        return image;
    }
};

File: Rust/Easy/TimeLimitExceeded/flood-fill_7.rs
impl Solution {
    pub fn flood_fill(image: Vec<Vec<i32>>, sr: i32, sc: i32, color: i32) -> Vec<Vec<i32>> {
        
        let mut image = image;
        let (sr, sc) = (sr as usize, sc as usize);
        let (m, n) = (image.len(), image[0].len());
        let color2 = image[sr][sc];
        let mut stack = vec![(sr, sc)];
        while !stack.is_empty() {
            let (i, j) = stack.pop().unwrap();
            if image[i][j] == color2 {
                image[i][j] = color;
                if i > 0 { stack.push((i-1, j)); }
                if i < m-1 { stack.push((i+1, j)); }
                if j > 0 { stack.push((i, j-1)); }
                if j < n-1 { stack.push((i, j+1)); }
            }
        }
        image
    }
}